relay-event-schema/src/protocol/span_v2.rs

```diff
@@ -0,0 +1,388 @@
+use opentelemetry_proto::tonic::trace::v1::{
+    span::SpanKind as OtelSpanKind, status::StatusCode as OtelSpanStatus,
+};
+use relay_protocol::{Annotated, Empty, FromValue, IntoValue, Object, Value};
+
+use std::fmt;
+use std::str::FromStr;
+
+use serde::Serialize;
+
+use crate::processor::ProcessValue;
+use crate::protocol::{Attribute, SpanId, Timestamp, TraceId};
+
+use super::OperationType;
+
+/// A version 2 (transactionless) span.
+#[derive(Clone, Debug, Default, PartialEq, Empty, FromValue, IntoValue)]
+pub struct SpanV2 {
+    #[metastructure(required = true, trim = false)]
+    pub trace_id: Annotated<TraceId>,
+
+    /// The ID of the span enclosing this span.
+    pub parent_span_id: Annotated<SpanId>,
+
+    /// The ID of the trace the span belongs to.
+    /// The Span id.
+    #[metastructure(required = false, trim = false)]
+    pub span_id: Annotated<SpanId>,
+
+    /// Span type (see `OperationType` docs).
+    #[metastructure(max_chars = 128)]
+    pub name: Annotated<OperationType>,
+
+    /// The span's status.
+    pub status: Annotated<SpanV2Status>,
+
+    /// Indicates whether a span's parent is remote.
+    ///
+    /// For OpenTelemetry spans, this is derived from span flags bits 8 and 9. See
+    /// `SPAN_FLAGS_CONTEXT_HAS_IS_REMOTE_MASK` and `SPAN_FLAGS_CONTEXT_IS_REMOTE_MASK`.
+    ///
+    /// The states are:
+    ///  - `empty`: unknown
+    ///  - `false`: is not remote
+    ///  - `true`: is remote
+    pub is_remote: Annotated<bool>,
+
+    // Used to clarify the relationship between parents and children, or to distinguish between
+    // spans, e.g. a `server` and `client` span with the same name.
+    //
+    // See <https://opentelemetry.io/docs/specs/otel/trace/api/#spankind>
+    #[metastructure(skip_serialization = "empty", trim = false)]
+    pub kind: Annotated<SpanV2Kind>,
+
+    /// Timestamp when the span started.
+    #[metastructure(required = true)]
+    pub start_timestamp: Annotated<Timestamp>,
+
+    /// Timestamp when the span was ended.
+    #[metastructure(required = true)]
+    pub end_timestamp: Annotated<Timestamp>,
+
+    /// Arbitrary attributes on a span.
+    #[metastructure(pii = "true", trim = false)]
+    pub attributes: Annotated<Object<Attribute>>,
+
+    /// Additional arbitrary fields for forwards compatibility.
+    #[metastructure(additional_properties, retain = true, pii = "maybe")]
+    pub other: Object<Value>,
+}
+
+impl SpanV2 {
+    pub fn attribute(&self, key: &str) -> Option<&Annotated<Value>> {
+        Some(&self.attributes.value()?.get(key)?.value()?.value.value)
+    }
+}
+
+/// Status of a V2 span.
+///
+/// This is a subset of OTEL's statuses (unset, ok, error).
+#[derive(Clone, Copy, Debug, PartialEq, Serialize)]
+#[serde(rename_all = "snake_case")]
+#[repr(i32)]
+pub enum SpanV2Status {
```

Comment:
The Notion doc says this type should be "a subset" of the OTEL span status, comprising only `"ok"` and `"error"`, but I opted for carrying over `"unset"` as well, similarly to how we do it in `SpanKind`.

Reply:
We do not really have a use-case for `unset` in our SDKs right now. Adding it later should be possible, but up to you.